webservices/cxf/src/java/org/sakaiproject/webservices/SakaiScript.java

```diff
@@ -4257,4 +4257,148 @@ private void setToolProperties(ToolConfiguration tool, String propList) {
+            for (iRoles = roles.iterator(); iRoles.hasNext(); ) {
+                Role r = (Role) iRoles.next();
+                Node item = dom.createElement("role");
+                Node roleId = dom.createElement("roleId");
+                roleId.appendChild(dom.createTextNode(r.getId()));
+                Node roleDescription = dom.createElement("roleDescription");
+                roleDescription.appendChild(dom.createTextNode(r.getDescription()));
+                item.appendChild(roleId);
+                item.appendChild(roleDescription);
+                list.appendChild(item);
+            }
+        } catch (Exception e) {
+            LOG.error("WS getAvailableRoles(): " + e.getClass().getName() + " : " + e.getMessage(), e);
+            return "";
+        }
+        return Xml.writeDocumentToString(dom);
+    }
+
+    @WebMethod
+    @Path("/getExistingFunctions")
+    @Produces("text/plain")
+    @GET
+    public String getExistingFunctions(
+            @WebParam(name = "sessionid", partName = "sessionid") @QueryParam("sessionid") String sessionId,
+            @WebParam(name = "authzgroupid", partName = "authzgroupid") @QueryParam("authzgroupid") String authzgroupid,
+            @WebParam(name = "roleid", partName = "roleid") @QueryParam("roleid") String roleid) {
+        Session session = establishSession(sessionId);
+
+        Document dom = Xml.createDocument();
+        Node list = dom.createElement("list");
+        dom.appendChild(list);
+
+        try {
+            AuthzGroup authzgroup = authzGroupService.getAuthzGroup(authzgroupid);
+            Role role = authzgroup.getRole(roleid);
+
+            //get functions that are in this role
+            Set existingfunctions = role.getAllowedFunctions();
+            Iterator it = existingfunctions.iterator();
+
+            Node item = dom.createElement("functions");
+            while (it.hasNext()) {
+                Node function = dom.createElement("function");
+                function.appendChild(dom.createTextNode((String) it.next()));
+                item.appendChild(function);
+            }
+            list.appendChild(item);
+        } catch (Exception e) {
+            LOG.error("WS getAvailableRoles(): " + e.getClass().getName() + " : " + e.getMessage(), e);
+            return "";
+        }
+        return Xml.writeDocumentToString(dom);
+    }
+
+    @WebMethod
+    @Path("/getSiteDefaultJoinerRole")
+    @Produces("text/plain")
+    @GET
+    public String getSiteDefaultJoinerRole(
+            @WebParam(name = "sessionid", partName = "sessionid") @QueryParam("sessionid") String sessionid,
+            @WebParam(name = "siteid", partName = "siteid") @QueryParam("siteid") String siteid) {
+
+        Session session = establishSession(sessionid);
+
+        try {
+            Site site = siteService.getSite(siteid);
+            if (site != null) {
+                return site.getJoinerRole();
+            } else {
+                return "failure";
```

Comment:
This should probably not return a String as the error when the OK response is also a String. Also it will look like an error if the joiner role is actually 'failure'. Throw an exception or log it and return nothing IMO.